xwayland/server.c

```diff
@@ -332,15 +345,24 @@ static bool server_start(struct wlr_xwayland_server *server) {
 		pid_t pid = fork();
 		if (pid < 0) {
 			wlr_log_errno(WLR_ERROR, "second fork failed");
+			(void) !write(p[1], "", 0);
```

Comment:
Actually, do we even need that write? If the parent receives `WL_EVENT_HANGUP`, it can detect the failure?

Reply:
Can you explain to me what triggers WL_EVENT_HANGUP? The write end of the pipe being closed? The general issue with that is that we might leak the write fd to another process, as creating the pipe and setting the fds CLOEXEC isn't atomic, so there's a race. If it weren't for that, none of the code paths would have to write, we could just close the fd. But since another process might keep the write end open, we have to actually do a write.

Reply:
I'd really like to be able to detect Xwayland exec failures. Xwayland might fail to exec if the executable is missing for instance.

>Can you explain to me what triggers WL_EVENT_HANGUP? The write end of the pipe being closed?

Yes.

>The general issue with that is that we might leak the write fd to another process, as creating the pipe and setting the fds CLOEXEC isn't atomic, so there's a race.

This can only happen if the compositor is multi-threaded and forks+execs outside of the wlroots thread. POSIX doesn't give us tools (yet) to atomically create a CLOEXEC pipe, so I'd just ignore the race for now.

Reply:
> I'd really like to be able to detect Xwayland exec failures. Xwayland might fail to exec if the executable is missing for instance.

But we are currently detecting that. 

- If the second fork (line 345) fails, the write to the pipe (line 348) wakes up `xserver_handle_ready`, which will waitpid on the first fork and see that it exited with EXIT_FAILURE
- If the exec fails, the write (line 356) will wake up `xserver_handle_ready`, but not before the `if (sig == SIGCHLD)` branch exited with EXIT_FAILURE

When there was no failure, the same write (line 356) wakes up the handler, which now sees EXIT_SUCCESS – which can only happen if the exec succeeded and Xwayland sent a SIGUSR1.

> This can only happen if the compositor is multi-threaded and forks+execs outside of the wlroots thread.

Sure, but it _can_ happen, and when it does happen, we'll get hangs. The write avoids that risk.

Reply:
>waitpid on the first fork and see that it exited with EXIT_FAILURE

Ah. I forgot we waitpid'ed the first fork. An issue with this behavior is compositors can't set a global SIGCHLD handler without stealing the Xwayland process status. But that's a completely separate issue.